grammar.js

```diff
@@ -1234,6 +1270,7 @@ module.exports = grammar({
     [$._expression, $.type_specifier],
     [$._expression, $.field_initializer],
     [$.scoped_identifier, $._type_constant],
+    [$.context_const_declaration,$._member_modifier],
```

Comment:
I don't know why this is necessary. Without it, the parsing is ambiguous (which I do not understand why, as `$._member_modifier` alone should never run in the same context as `$.context_const_declaration`.
The error I get without this line is:
```
Unresolved conflict for symbol sequence:

  'trait'  identifier  '{'  abstract_modifier  •  'const'  …

Possible interpretations:

  1:  'trait'  identifier  '{'  (_member_modifier  abstract_modifier)  •  'const'  …
  2:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  ';')
  3:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  '='  capability_list  ';')
  4:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  'as'  capability_list  ';')
  5:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  'as'  capability_list  '='  capability_list  ';')
  6:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  'super'  capability_list  ';')
  7:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  'super'  capability_list  '='  capability_list  ';')
  8:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  'super'  capability_list  'as'  capability_list  ';')
  9:  'trait'  identifier  '{'  (context_const_declaration  abstract_modifier  •  'const'  'ctx'  identifier  'super'  capability_list  'as'  capability_list  '='  capability_list  ';')

Possible resolutions:

  1:  Specify a higher precedence in `context_const_declaration` than in the other rules.
  2:  Specify a higher precedence in `_member_modifier` than in the other rules.
  3:  Specify a left or right associativity in `_member_modifier`
  4:  Add a conflict for these rules: `context_const_declaration`, `_member_modifier`
```
so to resolve it i took option 4.

Reply:
Gosh, that's a good question. I think tree-sitter is complaining that all of the following are legal in Hack:

```
abstract class Foo {
  abstract const int FOO;
  abstract const type TBaz;
  abstract const ctx CBar;
}
```

I see the parsing for `const ctx` was introduced in D25515839. It looks like `const type` and `const ctx` are checked first in the parser, so they take precedence. (This also means that you're gonna have a bad time if you define a type called `ctx`.)

Based on this, I suggest you take option 1 here.

Reply:
By the way, this currently parses perfectly using this method! (which is what confuses me)
For example, the parse tree for the example you gave shows up for me as
```
class_declaration [0, 0] - [4, 1]
  abstract_modifier [0, 0] - [0, 8]
  name: identifier [0, 15] - [0, 18]
  body: member_declarations [0, 19] - [4, 1]
    const_declaration [1, 2] - [1, 25]
      abstract_modifier [1, 2] - [1, 10]
      type: type_specifier [1, 17] - [1, 20]
      const_declarator [1, 21] - [1, 24]
        name: identifier [1, 21] - [1, 24]
    type_const_declaration [2, 2] - [2, 27]
      abstract_modifier [2, 2] - [2, 10]
      name: identifier [2, 22] - [2, 26]
    context_const_declaration [3, 2] - [3, 26]
      abstract_modifier [3, 2] - [3, 10]
      name: identifier [3, 21] - [3, 25]
```

I thought the same thing as you when I saw this issue for the first time, so I tried raising the precedence of `const ctx`, but I think since the parser doesn't backtrack this led to tons of failures in other tests, where it would prefer `const ctx` over other const declarations and fail.
I also tried adding right precedence to `_member_modifier` so it would bind to the const declaration, but this also caused failuse in `const ctx` parsing.

I think adding a conflit means (somewhat unintuitively) that it _will_ backtrack, and parse each subtree fully before making a decision, and since none of these should parse at the same time (of course assuming `ctx` isn't a type) then it will work out OK.
I think this is assuming it will try to favor `'ctx'` over any random identifier for type parsing.... but this is what I have seen in my testing. A way to get around this assumption might be to keep the conflict rule and set a dynamic precedence (but I haven't had the greatest luck with those either)

Reply:
Conflicts are (IMO) the most challenging part of making fast, accurate grammars. @4e554c4c you are correct that adding a conflict will make it backtrack after looking forwards (in tree-sitter's words: "Tree-sitter will use the GLR algorithm to explore all of the possible interpretations"). When possible, we try to avoid them, but they seem fine in this situation. Dynamic precedence could be used in tandem with the conflict, but seems a bit unnecessary here. 

One just must live with the fact that as a pseudo-LR (GLR) parser, tree-sitter can only handle context-free grammar. 🤷

Reply:
FWIW the Hack parser itself does do backtracking in a bunch of places, so it's probably correct for tree-sitter-hack to also have some backtracking.